compat/src/test/csharp/Program.cs

```diff
@@ -0,0 +1,399 @@
+            }
+
+            if (testFailed)
+            {
+                Environment.Exit(1);
+            }
+        }
+    }
+
+    private async Task ExecuteSteps(List<Step> steps, Dictionary<string, object> variables, NpgsqlConnection connection)
+    {
+        foreach (var step in steps)
+        {
+            if (step.Loop != null)
+            {
+                await ExecuteLoop(step.Loop, variables, connection);
+            }
+            else
+            {
+                await ExecuteStep(step, variables, connection);
+            }
+        }
+    }
+
+    private async Task ExecuteLoop(LoopDefinition loop, Dictionary<string, object> variables,
+        NpgsqlConnection connection)
+    {
+        var loopVariables = new Dictionary<string, object>(variables);
+
+        IEnumerable<object> iterable;
+        if (loop.Over != null)
+        {
+            iterable = loop.Over;
+        }
+        else if (loop.Range != null)
+        {
+            iterable = Enumerable.Range(loop.Range.Start, loop.Range.End - loop.Range.Start + 1).Cast<object>();
+        }
+        else
+        {
+            throw new ArgumentException("Loop must have 'over' or 'range' defined.");
+        }
+
+        foreach (var item in iterable)
+        {
+            loopVariables[loop.As] = item;
+            await ExecuteSteps(loop.Steps, loopVariables, connection);
+        }
+    }
+
+    private async Task ExecuteStep(Step step, Dictionary<string, object> variables, NpgsqlConnection connection)
+    {
+        var queryTemplate = step.Query;
+        var parameters = step.Parameters ?? new List<Parameter>();
+        var expect = step.Expect;
+
+        var substitutedQuery = SubstituteVariables(queryTemplate, variables);
+        var query = AdjustPlaceholderSyntax(substitutedQuery);
+
+        var resolvedParameters = ResolveParameters(parameters, variables);
+        var result = await ExecuteQuery(connection, query, resolvedParameters);
+
+        if (expect != null)
+        {
+            AssertResult(expect, result);
+        }
+    }
+
+    private string SubstituteVariables(string template, Dictionary<string, object> variables)
+    {
+        if (string.IsNullOrEmpty(template)) return template;
+
+        return variables.Aggregate(template, (current, variable) =>
+            current.Replace($"${{{variable.Key}}}", variable.Value.ToString() ?? ""));
+    }
+
+    private string AdjustPlaceholderSyntax(string query)
+    {
+        return System.Text.RegularExpressions.Regex.Replace(query, @"\$\[(\d+)\]", "@p$1");
+    }
+
+    private async Task<object> ExecuteQuery(NpgsqlConnection connection, string query, List<NpgsqlParameter> parameters)
+    {
+        var queryType = query.Trim().Split(' ')[0].ToLower();
+
+        await using var cmd = new NpgsqlCommand(query, connection);
+        cmd.Parameters.AddRange(parameters.ToArray());
+
+        if (queryType == "select")
```

Comment:
What about queries like `my_table LIMIT -1`? Is there a guarantee this syntax will never be used?

Reply:
there is no guarantee. I don't know how to check this. the compiler does not necessarily have access to the actual values since SQLAlchemy allows binding variables.